dbms/src/Functions/FunctionsGeo.cpp

```diff
@@ -246,6 +251,181 @@ class FunctionPointInPolygon : public IFunction
 
 };
 
+const size_t GEOHASH_MAX_TEXT_LENGTH = 16;
+
+// geohashEncode(lat float32/64, lon float32/64, length UInt8) => string
+class FunctionGeohashEncode : public IFunction
+{
+public:
+    static constexpr auto name = "geohashEncode";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionGeohashEncode>(); }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override { return 3; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        validateArgumentType(*this, arguments, 0, isFloat, "float");
+        validateArgumentType(*this, arguments, 1, isFloat, "float");
+        validateArgumentType(*this, arguments, 2, isInteger, "integer");
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    template <typename PrecisionType, typename LatColumnType, typename LonColumnType>
+    bool tryExecute2(const LatColumnType * lat_column, const LonColumnType * lon_column, const IColumn * precision_col, ColumnPtr & result)
+    {
+        const ColumnVector<PrecisionType> * precision_column = checkAndGetColumnEvenIfConst<ColumnVector<PrecisionType>>(precision_col);
+        if (!precision_column)
+            return false;
+
+        auto col_str = ColumnString::create();
+        ColumnString::Chars & out_vec = col_str->getChars();
+        ColumnString::Offsets & out_offsets = col_str->getOffsets();
+
+
+        const size_t size = lat_column->size();
+
+        out_offsets.resize(size);
+        out_vec.resize(size * (GEOHASH_MAX_TEXT_LENGTH + 1));
+
+        char * begin = reinterpret_cast<char *>(out_vec.data());
+        char * pos = begin;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            const Float64 latitude = lat_column->getElement(i);
+            const Float64 longitude = lon_column->getElement(i);
+            const Int64 precision = precision_column->getElement(i);
```

Comment:
It doesn't work for const column. For example, the following query will throw an exception:

```
create table t(x Float32, y Float32) engine=MergeTree order by tuple();
insert into t values (1.,2.), (3.,4.), (5.,6.), (7.,8.),(9., 10.), (11.,12.)
select geohashEncode(x, y, 3) from t
```

When `precision_column` is a const column it actually contains a single value (not `size` of values) of so `precision_column->getElement(i)` isn't correct for this case.

Reply:
I suppose you can fix it up by marking the `precision_column` as always-constant:

```
virtual ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {2}; }
```

and then by using
```
const Float64 latitude = lat_column->getElement(i);
const Float64 longitude = lon_column->getElement(i);
const Int64 precision = precision_column->getElement(0);
```
while iterating.

Reply:
Nice catch about mixing const/non-const columns! Lemme fix it :)

Reply:
Fixed